docs/source/advanced_usage.rst

```diff
@@ -32,6 +32,87 @@ locally):
 - QVM URL: ``http://127.0.0.1:5000``
 - quilc URL: ``tcp://127.0.0.1:5555``
 
+Concurrency
+~~~~~~~~~~~
+
+Using pyQuil for concurrent programming is as simple as calling :py:func:`~pyquil.get_qc()` from within a given context
+(be it a thread or process), then using the returned :py:class:`~pyquil.api.QuantumComputer` as usual within that
+context. While :py:class:`~pyquil.api.QuantumComputer` objects as a whole are not safe to share between contexts (due to
+state related to currently-running compilation or execution requests), some information they use is. Information related
+to client configuration (:py:class:`~pyquil.api.QCSClientConfiguration`) and QPU auth
+(:py:class:`~pyquil.api.EngagementManager`) can be safely extracted and shared among
+:py:class:`~pyquil.api.QuantumComputer` instances, as shown below, to save your code from redundant disk reads and
+auth-related HTTP requests.
+
+.. note::
+    QVMs process incoming requests in parallel, while QPUs process them sequentially. If you encounter timeouts while
+    trying to run large numbers of programs against a QPU, try increasing the ``timeout`` parameter on calls to
+    :py:func:`~pyquil.get_qc()` (specified in seconds).
+
+Using Multithreading
+--------------------
+
+.. code:: python
+
+    from multiprocessing.pool import ThreadPool
+
+    from pyquil import get_qc, Program
+    from pyquil.api import EngagementManager, QCSClientConfiguration
+
+
+    configuration = QCSClientConfiguration.load()
+    engagement_manager = EngagementManager(client_configuration=configuration)
+
+
+    def run(program: Program):
+        qc = get_qc("Aspen-8", client_configuration=configuration, engagement_manager=engagement_manager)
+        return qc.run(qc.compile(program))
+
+
+    programs = [Program("DECLARE ro BIT", "RX(pi) 0", "MEASURE 0 ro").wrap_in_numshots_loop(10)] * 20
+    with ThreadPool(5) as pool:
+        results = pool.map(run, programs)
+
+    for i, result in enumerate(results):
+        print(f"Results for program {i}:\n{result}\n")
+
+
+Using Multiprocessing
+---------------------
+
+.. code:: python
+
+    from multiprocessing.pool import Pool
+
+    from pyquil import get_qc, Program
+    from pyquil.api import EngagementManager, QCSClientConfiguration
+
+
+    configuration = QCSClientConfiguration.load()
```

Comment:
I would leave a comment as to why initializing `configuration` and `engagement_manager` outside of the `run` def is important.

To that point, I believe this requires an update in qcs-api-client-python to make sure updates are threadsafe https://gitlab.com/rigetti/qcs/clients/qcs-api-client-python/-/issues/15

Reply:
In the paragraph above the code examples, it states the reason. Let me know if this is still unclear though:

> Information related
to client configuration (:py:class:`~pyquil.api.QCSClientConfiguration`) and QPU auth
(:py:class:`~pyquil.api.EngagementManager`) can be safely extracted and shared among
:py:class:`~pyquil.api.QuantumComputer` instances, as shown below, to save your code from redundant disk reads and
auth-related HTTP requests.

As far as updating the configuration, I was assuming that 99% of use cases are going to only be reading config and not updating it, so this optimization is more about saving redundant loads of the config. That's still a good catch though, and probably something worth addressing at some point. Thanks for creating the issue :)